### ISSUE TITLE
Is there value in defining an integrity scheme?

### ISSUE THREAD

**Issue by mjoras** [nibanks/quic-disable-encryption#2]
> Because the AEAD tag is removed along with the encryption, the UDP checksum must be relied upon to determine any packet corruption.

The Internet checksum is known to have quite weak guarantees. This is mitigated by the fact that the underlying (presumably Ethernet in most cases) link layer has better checksumming, but there are still cases where it may be useful to have a protocol-level integrity check. One such example is a faulty NIC w/ checksum offloading which may be providing "verified" UDP payloads to the OS stack.

Given that the cost of some form of integrity tag (e.g. a CRC32) is vanishingly small, should this extension provide some way to add one?

**Comment by nibanks** [comment]
It's a valid concern, but the question I have then, is for scenarios that would actually use this, when would you expect the UDP checksum not to be enough. This is either for performance testing, tunneled environments or walled gardens. Most of these I think wouldn't really need the extra integrity checks. What do you think?

**Comment by mjoras** [comment]
It's a good question. The situation I was specifically thinking of was internally we had an incident where one of our load balancers had a NIC which had UDP checksum offloading enabled (which we enable globally), but the NIC itself was faulty, resulting in the network stack believing that the UDP checksum was valid (and not checking it), while the NIC passed up corrupted packets. The QUIC packets in this situation were obviously dropped and we noticed a big spike in failures to decrypt. If we had instead been running with this extension, the data could have been silently corrupted.

This is a somewhat esoteric situation, I'll admit, but it is one that can happen at scale in modern DCs. And one solution potentially would be to always disable UDP checksum offloading.

**Comment by nibanks** [comment]
I totally agree something like this can happen. I've seem similar issues with TCP too though. From my experience, when this happens, it just because the next layer up (the app protocol) that ends up identifying the problem (not the transport). Personally, I'd really like to just say that is a bug, and you shouldn't need to design for bugs. But I agree that's not a completely realistic idea.

Also, if we add an integrity check here, there's 0% chance we're ever going to get HW offload support for it. So it's a perf bottleneck (albeit small) that would be pretty much set in stone.
